### DISCUSSION TITLE
All about the menu(s) topic

### DISCUSSION THREAD

**Discussion by wormstein** [EdNekebno/DGTCentaurMods#37] [General]
Hi,
I was thinking about the menu and I am sure we can restructure it so it'll be easy to extend.

The way I view this is like so: have a "menu" folder where we can have a module for eacch menu option. For example centaur-menu.py or lichess-menu.py where we can build all the menu logic and submenus. Entry point can be main-menu.py where we simply add or remove the others. This way we avoid menu.py to become huge... When we need to change one, we just edit the corresponding file.

Curreent menu.py can become main.py and should hold anything related to board startup, like any threads, scripts, hardware, etc...

I will add this in the project board. I just want to hear your other thoughts on this.

**Comment by EdNekebno** [comment]
Sounds interesting, some thoughts...

1. The first thing that comes to mind is that the code for the menu is a complete pain :) I spent several hours simply speeding it up a little bit at the weekend. The reason for this is that it has to keep track of the display state, etc, to avoid unnecessary refreshes (which are slow).

2. There is a doMenu function in boardfunctions.py . This displays a single level menu. It's called, for example, also in my wifi code just to list scanned SSIDs. menu.py calls this and essentially holds the submenu structure and actions by calling doMenu repeatedly.

3. I have been thinking it might be better to have a menu function that knows the entire structure in advanced (including submenus). It makes the tracking of the display state a little easier as well as calculating what part of the screen to clear and redraw when changing.

4. Some submenus should actually probably be dynamic. e.g. for Lichess it currently just has a submenu of "Current" or "New Game". But then "New Game" should list time controls and "Current" should dynamically fetch games from Lichess (and therefore probably be controlled from that point by lichess.py).

The idea, I think, breaks (3) but that's not a deal breaker for me. An alternative option, I think, is that when we have a database setup to then define the menus in there. That has a major advantage that it can eventually be a user configurable option if something displays, the order of display, etc. Given e-paper speed, it's a big advantage to users to be able to do that (if someone never uses Lichess they can hide it and save themselves a button press).

**Comment by wormstein** [comment]
I agree here. Centaur software has a drawMenu function too which I think works the same way.
We can go this approach. Have menu rendering functions in a menu.py which will just take arguments. This will fit nice when database will come into play... 

Let's separate first the menu part from the menu.py somehow and rename menu.py to main.py so it's the entry point of our mods.

**Comment by wormstein** [comment]
Do you know what is the button index for the question button? I'm too lazy to search it, I know :)
```
    if (buttonPress == 1):
             epd.Clear(0xff)
             return "BACK"
```
I would like to add the option in boardfunction when it is pressed to show the QR...

**Reply by EdNekebno** [reply]
BTNBACK = 1
BTNTICK = 2
BTNUP = 3
BTNDOWN = 4
BTNHELP = 5
BTNPLAY = 6

I also added them to boardfunctions when I made the last changes. You want BTNHELP,  so 5.

**Comment by wormstein** [comment]
In case you wonder what the 'Restore' option in wifi menu is, I am building a function to be able to regain network access to the board  in case something goes wrong. It will simply restore a backup of wpa_supplicant conf file made when network was confirmed as working...
No need to remove the RPi to access the sd.

**Reply by EdNekebno** [reply]
Great idea. You seem to understand my potential to muck up my settings on anything :)

**Reply by wormstein** [reply]
:)) Finishing those features for wifi is a very fragile thing to test...
